### PR TITLE
Sanitize some wallet serialization

### PR DESCRIPTION
**Pull request by sipa**
This is a small subset of changes taken from #10785, fixing a few of the craziest constness violations in the serialization code.

`CWalletTx` currently serializes some of its fields by embedding them in a key-value `mapValue`, which is modified (and then fixed up) even from the `Serialize` method (for which `mapValue` is const). `CAccountingEntry` goes even further in that it stores such a map by appending it into `strComment` after a null char, which is again later fixed up again.

Fix this by splitting the serialization and deserialization code, and making the serialization act on a copy of `mapValue` / `strComment`.

### CODE REVIEW

File: src/wallet/wallet.h [hunk 1]
```diff
@@ -608,48 +602,49 @@ class CAccountingEntry
...
-                ss.insert(ss.end(), _ssExtra.begin(), _ssExtra.end());
-                strComment.append(ss.str());
-            }
+        s << nCreditDebit << nTime << strOtherAccount;
+
+        mapValue_t mapValueCopy = mapValue;
+        WriteOrderPos(nOrderPos, mapValueCopy);
+
+        std::string strCommentCopy = strComment;
+        if (!(mapValueCopy.empty() && _ssExtra.empty())) {
```
Comment (eklitzke) [inline src/wallet/wallet.h]: nit: consider making this:

```c++
if (!mapValueCopy.empty() || !_ssExtra.empty()) {
```

**Review by eklitzke — COMMENTED** [review]
I'm not an expert in this part of the code, but this change looks good, and I'm happy to see so many red lines in the diff.

File: src/wallet/wallet.h [hunk 2]
```diff
@@ -390,42 +390,36 @@ class CWalletTx : public CMerkleTx
...
+        Init(nullptr);
+        char fSpent;
 
-            nTimeSmart = mapValue.count("timesmart") ? (unsigned int)atoi64(mapValue["timesmart"]) : 0;
-        }
+        s >> *static_cast<CMerkleTx*>(this);
+        std::vector<CMerkleTx> vUnused; //!< Used to be vtxPrev
+        s >> vUnused >> mapValue >> vOrderForm >> fTimeReceivedIsTxTime >> nTimeReceived >> fFromMe >> fSpent;
+
+        strFromAccount = std::move(mapValue["fromaccount"]);
```
[the remark this answers is not in the document]
Comment (sipa) [reply src/wallet/wallet.h]: As this code isn't particularly performance critical, I prefer the current shorter code.

File: src/wallet/wallet.h (hunk 1 above)
Comment (sipa) [reply src/wallet/wallet.h]: Done.

File: src/wallet/wallet.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (eklitzke) [reply src/wallet/wallet.h]: Makes sense.

**Comment by promag** [comment]
utACK 42343c7 LGTM.

**Review by ryanofsky — COMMENTED** [review]
utACK 42343c748c2bca8ba9b888d949088dbfd1f142b4. Nice to make  #10785 a little simpler.

File: src/wallet/wallet.h [hunk 3]
```diff
@@ -608,48 +602,49 @@ class CAccountingEntry
...
         int nVersion = s.GetVersion();
-        if (!(s.GetType() & SER_GETHASH))
-            READWRITE(nVersion);
+        if (!(s.GetType() & SER_GETHASH)) {
+            s << nVersion;
+        }
         //! Note: strAccount is serialized as part of the key, not here.
-        READWRITE(nCreditDebit);
-        READWRITE(nTime);
-        READWRITE(LIMITED_STRING(strOtherAccount, 65536));
```
[the remark this answers is not in the document]
Comment (laanwj) [reply src/wallet/wallet.h]: Although you could add a sanity check, serialization is probably too late to check it. Ideally anything inputting account/label names should check this to be able to give an useful error and reject setting it before putting it in a data structure.
But yes, out of scope here.

**Comment by laanwj** [comment]
utACK 42343c748c2bca8ba9b888d949088dbfd1f142b4

**Pull request merged**